specs/WebNotification.md

```diff
@@ -0,0 +1,610 @@
+                               : L"This request did not come from a user gesture.");
+
+            int response = MessageBox(
+                nullptr, message.c_str(), L"Permission Request",
+                MB_YESNOCANCEL | MB_ICONWARNING);
+            switch (response)
+            {
+            case IDYES:
+                m_cached_permissions[cached_key] = true;
+                state = COREWEBVIEW2_PERMISSION_STATE_ALLOW;
+                break;
+            case IDNO:
+                m_cached_permissions[cached_key] = false;
+                state = COREWEBVIEW2_PERMISSION_STATE_DENY;
+                break;
+            default:
+                state = COREWEBVIEW2_PERMISSION_STATE_DEFAULT;
+                break;
+            }
+        }
+
+        CHECK_FAILURE(args->put_State(state));
+        CHECK_FAILURE(deferral->Complete());
+    });
+    return S_OK;
+}
+
+```
+
+## Filter Notifications from a specific doamin and send local toast
+
+Learn more about sending a local toast [Send a local toast notification from a C# app - Windows apps | Microsoft Learn](https://learn.microsoft.com/windows/apps/design/shell/tiles-and-notifications/send-local-toast?tabs=uwp).
+
+### C# Sample
+```csharp
+using Microsoft.Toolkit.Uwp.Notifications;
+...
+void WebView_NotificationReceived(object sender, CoreWebView2NotificationReceivedEventArgs args)
+{
+    CoreWebView2Deferral deferral = args.GetDeferral();
+    var notification = args.Notification;
+
+    // Show notification with local MessageBox
+    string message = "Received notification from " + args.Uri + " with body " + notification.Body;
+    MessageBox.Show(message);
+
+    // Requires Microsoft.Toolkit.Uwp.Notifications NuGet package version 7.0 or greater
+    var toastContent = new ToastContentBuilder()
+        .AddArgument("conversationId", m_id)
+        .AddText("Notification sent from " + args.Uri +":\n")
+        .AddText(notification.Body);
+    toastContent.Show();
+    // See
+    // https://learn.microsoft.com/windows/apps/design/shell/tiles-and-notifications/send-local-toast?tabs=uwp
+    // for how to handle toast activation
+
+    notification.Show();
+    notification.Close();
+    args.Handled = true;
+}
+```
+
+### C++ Sample
+```cpp
+...
+{
+    //! [NotificationReceived]
+    // Register a handler for the NotificationReceived event.
+    CHECK_FAILURE(m_webView2_17->add_NotificationReceived(
+        Callback<ICoreWebView2NotificationReceivedEventHandler>(
+            [this](
+                ICoreWebView2* sender,
+                ICoreWebView2NotificationReceivedEventArgs* args) -> HRESULT
+            {
+                // Block notifications from specific URIs and set Handled to
+                // false so the the default notification UI will not be
+                // shown by WebView2 either.
+                wil::unique_cotaskmem_string uri;
+                CHECK_FAILURE(args->get_Uri(&uri);
+                if (ShouldBlockUri(uri.get()))
```

Comment:
This is confusing to me. If we want to block the uri from showing notification, shouldn't we mark it handled and then not do anything else? 

For default UI it would be like the following?

eventArgs.handled = shouldBlock(notification);

Or if we're doing custom UI then its like

eventArgs.handled = true;
if (!shouldBlock(notification)) {
showNotification(notification)
}

Is that right or do I misunderstand the API

Reply:
Your interpretation is correct. - we should mark it as **handled** to stop default UI from showing if shouldBlock. I had a typo and was mistaken. Will fix it.